ui/test/nuxeo-document-layout.test.js

```diff
@@ -0,0 +1,232 @@
+import '../nuxeo-document-layout.js';
+import '../widgets/nuxeo-input.js';
+import '../widgets/nuxeo-textarea';
+import { Polymer } from '@polymer/polymer/polymer-legacy.js';
+import {
+  fixture,
+  flush,
+  isElementVisible,
+  html,
+  waitForEvent,
+  waitForAttrMutation,
+} from '@nuxeo/nuxeo-elements/test/test-helpers';
+import { LayoutBehavior } from '../nuxeo-layout-behavior.js';
+
+// Export Polymer and PolymerElement for 1.x and 2.x compat
+window.Polymer = Polymer;
+
+window.nuxeo.I18n.language = 'en';
+window.nuxeo.I18n.en = window.nuxeo.I18n.en || {};
+
+Nuxeo = Nuxeo || {};
+Nuxeo.LayoutBehavior = LayoutBehavior;
+
+/* eslint-disable no-unused-expressions */
+suite('<nuxeo-document-layout>', () => {
+  let documentLayout;
+
+  const buildDoc = (props) => {
+    const doc = { type: 'Test', uid: '12ae', properties: {} };
+    Object.assign(doc.properties, props);
+    return doc;
+  };
+
+  const buildLayout = async (doc = buildDoc(), layout = 'edit') => {
+    const dl = await fixture(
+      html`
+        <nuxeo-document-layout .document="${doc}" layout="${layout}"></nuxeo-document-layout>
+      `,
+    );
+    if (!dl.element) {
+      await Promise.race([
+        waitForEvent(dl.$.layout, 'element-changed'),
+        waitForAttrMutation(dl.$.layout.$.error, 'hidden', null),
+      ]);
+    }
+    await flush();
+    return dl;
+  };
+
+  const getWidget = (label) => {
+    if (!documentLayout || !documentLayout.element) {
+      return null;
+    }
+    return documentLayout.element.shadowRoot.querySelector(`[role="widget"][label="${label}"]`);
+  };
+
+  test('Should display an error when the layout is not found', async () => {
+    documentLayout = await buildLayout(buildDoc(), 'view'); // we have no such view layout, so it will result in 404
+    expect(documentLayout.$.layout.$.error.hidden).to.be.false;
+    expect(documentLayout.$.layout.$.error.message).to.equal('documentView.layoutNotFound');
+    isElementVisible(documentLayout.$.layout.$.error);
+    expect(documentLayout.element).to.be.undefined;
+  });
+
+  test('Should load a layout', async () => {
+    documentLayout = await buildLayout();
+    // check the layout was loaded
+    expect(documentLayout.element).to.not.be.null;
+    expect(documentLayout.element.tagName).to.equal('NUXEO-TEST-EDIT-LAYOUT');
+    // assert layout integrity
+    expect(documentLayout.element.document).to.equal(documentLayout.document);
+    const { children } = documentLayout.element.shadowRoot;
+    expect(children).to.have.lengthOf(3);
+    expect(children[0].tagName).to.equal('STYLE');
+    expect(children[1].tagName).to.equal('NUXEO-INPUT');
+    expect(children[2].tagName).to.equal('NUXEO-TEXTAREA');
+  });
+
+  test('Should fire a "document-layout-changed" event when the layout is loaded', async () => {
+    const ePromise = new Promise((resolve) => {
+      document.addEventListener('document-layout-changed', (e) => resolve(e));
+    });
```

Comment:
Probably we can use `waitForEvent`, no?

Reply:
So, I finally recall why it has to be done this way: the next line we does a `buildLayout()` which already calls flush, so we take the risk of not catching the error. However, in the other tests where this is used, we can replace this logic by a waitForEvent.